go/service/gregor.go

```diff
@@ -1169,21 +1171,46 @@ func (g *gregorHandler) auth(ctx context.Context, cli rpc.GenericClient) (err er
 }
 
 func (g *gregorHandler) pingLoop() {
+
+	duration := g.G().Env.GetGregorPingInterval()
+	timeout := g.G().Env.GetGregorPingTimeout()
+	g.Debug("ping loop: starting up: duration: %v timeout: %v", duration, timeout)
+	defer g.Debug("ping loop: terminating")
+
 	for {
 		select {
-		case <-g.G().Clock().After(g.G().Env.GetGregorPingInterval()):
-			_, err := gregor1.IncomingClient{Cli: g.cli}.Ping(context.Background())
+		case <-g.G().Clock().After(duration):
+
+			if !g.IsConnected() {
+				g.Debug("ping loop: skipping ping since not connected")
+				continue
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			_, err := gregor1.IncomingClient{Cli: g.cli}.Ping(ctx)
+			cancel()
```

Comment:
Well we don't normally return after calling this guy, so I thought it would be better to just call `cancel` right here, especially since we are done with the context at this point. I think most examples assume you aren't done with the context until you leave the function, but in this case our use is so narrow we can just run it here.